projects/roscoe/score.py

```diff
@@ -0,0 +1,1048 @@
+            return sum(grammar_scores) / len(grammar_scores)
+
+        grammar_results = self.check_grammar(hypo.chain, self.grammar_batch)
+        assert all(
+            hypo_step == grmr_step
+            for hypo_step, (grmr_step, _prob) in zip(hypo.chain, grammar_results)
+        ), "steps were returned out of order"
+
+        if GRAMMAR_STEP in score_types:
+            scores[GRAMMAR_STEP] = _average_results(grammar_results)
+        if GRAMMAR_STEP_MAX in score_types:
+            # Minimum correctness score, i.e. maximum probability of error
+            scores[GRAMMAR_STEP_MAX] = min(prob for _, prob in grammar_results)
+
+        return scores
+
+    def evaluate(self, score_types: Optional[List[str]] = None) -> Dict[str, Any]:
+        """
+        Score each reasoning chain.
+        """
+        score_types = score_types if score_types is not None else self.score_types
+        # scores[score_type] = [score_chain1, score_chain2, ...]
+        scores = defaultdict(list)
+
+        if contains_embedding_scores(score_types):
+            # Pre-compute chain, step, and word embeddings for embedding-based metrics
+            self.precompute_embeddings(
+                hypos=self.hypos,
+                contexts=self.context,
+                references=self.references,
+            )
+
+        for i in tqdm(range(len(self.hypos)), desc="Scoring chains ... "):
+            if len(self.hypos[i].chain) == 0:
+                for score in score_types:
+                    scores[score].append("N/A")
+                continue
+
+            if contains_embedding_scores(score_types):
+                # Calculate all embedding-based scores
+                scores = self.compute_embedding_scores(
+                    hypo=self.hypos[i],
+                    context=self.context[i],
+                    reference=self.references[i] if self.references else None,
+                    score_types=score_types,
+                    scores=scores,
+                )
+
+            # NLI model scores
+            if DISCOURSE_REPRESENTATION in score_types:
+                scores[DISCOURSE_REPRESENTATION].append(
+                    self.discourse_represenation(
+                        self.context[i].chain, self.hypos[i].chain, self.discourse_batch
+                    )
+                )
+            if any(s in score_types for s in (COHERENCE_STEP, COHERENCE_MAX)):
+                step_contradict_probs = self.contradiction_steps(
+                    self.hypos[i].chain, self.coherence_batch
+                )
+            if COHERENCE_STEP in score_types:
+                # sum_k [ nli(y_k, y_[1, .., k-1)] ] / K-1
+                scores[COHERENCE_STEP].append(
+                    1 - sum(step_contradict_probs) / len(step_contradict_probs)
+                )
+            if COHERENCE_MAX in score_types:
+                scores[COHERENCE_MAX].append(1 - max(step_contradict_probs))
```

Comment:
Should we remove `coherence_step` and `coherence_max` since `coherence_step_by_step` outperformed them and they weren't included in the paper? Perhaps we could also rename `coherence_step_by_step` to `self_consistency` to match the paper.